### DISCUSSION TITLE
Test-Driven Development By Example | George Aristy

### DISCUSSION THREAD

**Discussion by giscus** [llorllale/llorllale.github.io#60] [Announcements]
# Test-Driven Development By Example | George Aristy

Written by Kent Beck, Test-Driven Development By Example explains how TDD works by way of 2 examples: first building piece by piece a toy Money class, then the initial scaffolds of xUnit. The third and final part of the book walks the reader through several TDD patterns as well as refactoring patterns.

https://georgearisty.dev/posts/test-driven-development-by-example/

**Comment by opack** [comment]
Hi!

Very interesting article, thanks! :pray:

I was particularly interested in the chapter about [triangluation](https://georgearisty.dev/posts/test-driven-development-by-example/#triangulate) because I am using randomized values in my tests precisely to avoid triangulation, which seems a bit like a waste of time.

What you explain is very interesting, because I prefered randomized values as I thought that using one or two or even three differents values do not "prove" that my system is robust. But your insight on the refactoring phase that should also avoid duplication between the test and the implementation is the key.

That being said, I still find it a bit useless :sweat_smile: I mean: using this technique, one has to write multiple similar tests (at least 2), an implementation that will evolve, and then do a refactoring to all of this to remove all tests but one.
The remaining test will not express the complexity of the final solution, as there is no direct link between it and the implementation. It does not endorse its role of "documenting" the implementation. It cannot be used to lead to the implementation once again.

With a randomized value, on the other hand, I directly have to adjust my implementation to the fact that the value is not known. I cannot use a simple constant to make the test pass. I do not have to write multiple tests that will be removed. I do not have to be rigorous enough to not forget to do this test-implementation de-duplication refactoring. And my test do express the correct need served by the implementation (i.e.: given "any" two numbers, I return the sum).

You started your chapter saying that you would use randomization, so I understand that you explained the triangulation for the sake of pedagogy. But, over your experience and after reading this book, do you see any advantage of triangulation over the use of randomized values?

Thanks again for this article, and your response! ;-)
